src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1044 @@
+
+    // do as much parameter validation as we can
+    if (!req.newLocations.IsNull())
+    {
+        auto locationIter = req.newLocations.Value().begin();
+        while (locationIter.Next())
+        {
+            uint32_t aSelectedLocation = locationIter.GetValue();
+
+            // each item in this list SHALL match the LocationID field of an entry on the SupportedLocations attribute's list
+            // If the Status field is set to UnsupportedLocation, the StatusText field SHALL be an empty string.
+            VerifyOrExit(IsSupportedLocation(aSelectedLocation), locationStatus = SelectLocationsStatus::kInvalidSet;
+                         useLocationStatusText = true;
+                         ChipLogError(Zcl, "HandleSelectLocationsCmd - unsupported location %u", aSelectedLocation));
+
+            // todo Should we take the responsibility for checking that there are no duplicate locations
+            //  from the delegate's IsValidSelectLocationsSet method?
+
+            // check to see if parameter list and attribute still match
+            if (matchesCurrentSelectedLocations)
+            {
+                if (!mDelegate->GetSelectedLocationByIndex(listIndex, oldSelectedLocation) ||
+                    (aSelectedLocation != oldSelectedLocation))
+                {
+                    matchesCurrentSelectedLocations = false;
+                }
+            }
+        }
+
+        // after iterating with Next through DecodableType - check for failure
+        VerifyOrExit(CHIP_NO_ERROR == locationIter.GetStatus(), locationStatus = SelectLocationsStatus::kInvalidSet;
+                     strncat(locationStatusText, "SelectLocations command - newLocations parameter value decoding failed",
+                             kMaxSizeStatusText);
+                     ChipLogError(Zcl, "%s", locationStatusText));
+    }
+
+    // If the NewLocations field is the same as the value of the SelectedLocations attribute
+    // the SelectLocationsResponse command SHALL have the Status field set to Success and
+    // the StatusText field MAY be supplied with a human-readable string or include an empty string.
+    VerifyOrExit(!matchesCurrentSelectedLocations, cmdStatus = Status::Success);
+
+    // If the current state of the device doesn't allow for the locations to be selected,
+    // the SelectLocationsResponse command SHALL have the Status field set to InvalidInMode.
+    // if the Status field is set to InvalidInMode, the StatusText field SHOULD indicate why the request is not allowed,
+    // given the current mode of the device, which may involve other clusters.
+    // (note - locationStatusText to be filled out by delegated function for if return value is false)
+    VerifyOrExit(mDelegate->IsSetSelectedLocationsAllowed(locationStatusText),
+                 locationStatus = SelectLocationsStatus::kInvalidInMode;
+                 cmdStatus      = Status::Failure);
```

Comment:
@johnfierke Why are we returning Failure here? I think this should only be set if there is an error in processing the command such that we cannot send the response command.